bundles/org.openhab.binding.monopriceaudio/src/main/java/org/openhab/binding/monopriceaudio/internal/handler/MonopriceAudioHandler.java

```diff
@@ -0,0 +1,723 @@
+
+                    Stream<String> zoneStream = MonopriceAudioZone.VALID_ZONES.stream();
+
+                    // poll each zone up to the number of zones specified in the configuration
+                    zoneStream.limit(numZones).forEach((zoneName) -> {
+                        try {
+                            connector.queryZone(MonopriceAudioZone.valueOf(zoneName));
+                        } catch (MonopriceAudioException e) {
+                            logger.warn("Polling error: {}", e.getMessage());
+                        }
+                    });
+
+                    // if the last successful polling update was more than 2.25 intervals ago, the controller
+                    // is either switched off or not responding even though the connection is still good
+                    if ((System.currentTimeMillis() - lastPollingUpdate) > (pollingInterval * 2.25 * 1000)) {
+                        logger.warn("Controller not responding to status requests");
+                        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                                "Controller not responding to status requests");
+                        closeConnection();
+                        scheduleReconnectJob();
+                    }
+                }
+            }
+        }, INITIAL_POLLING_DELAY, pollingInterval, TimeUnit.SECONDS);
+    }
+
+    /**
+     * Cancel the polling job
+     */
+    private void cancelPollingJob() {
+        ScheduledFuture<?> pollingJob = this.pollingJob;
+        if (pollingJob != null) {
+            pollingJob.cancel(true);
+            this.pollingJob = null;
+        }
+    }
+
+    /**
+     * Update the state of a channel
+     *
+     * @param channel the channel
+     */
+    private void updateChannelState(MonopriceAudioZone zone, String channelType, MonopriceAudioZoneDTO zoneData) {
+        String channel = zone.name().toLowerCase() + CHANNEL_DELIMIT + channelType;
+
+        if (!isLinked(channel)) {
+            return;
+        }
+
+        State state = UnDefType.UNDEF;
+        switch (channelType) {
+            case CHANNEL_TYPE_POWER:
+                state = zoneData.isPowerOn() ? OnOffType.ON : OnOffType.OFF;
+                break;
+            case CHANNEL_TYPE_SOURCE:
+                state = new DecimalType(zoneData.getSource());
+                break;
+            case CHANNEL_TYPE_VOLUME:
+                long volumePct = Math.round((double) (Integer.parseInt(zoneData.getVolume()) - MIN_VOLUME)
+                        / (double) (MAX_VOLUME - MIN_VOLUME) * 100.0);
+                state = new PercentType(BigDecimal.valueOf(volumePct));
+                break;
+            case CHANNEL_TYPE_MUTE:
+                state = zoneData.isMuted() ? OnOffType.ON : OnOffType.OFF;
+                break;
+            case CHANNEL_TYPE_TREBLE:
+                state = new DecimalType(BigDecimal.valueOf(Integer.parseInt(zoneData.getTreble()) - TONE_OFFSET));
```

Comment:
Is there any change that zoneData.getTreble() wouldn't return a parsable number? If so then you should catch the exception, if not then I suggest changing the datatype from a `String` to an `int`.
```suggestion
                state = new DecimalType(BigDecimal.valueOf(Integer.parseInt(zoneData.getTreble()) - TONE_OFFSET));
```

Reply:
The regex ensures that only digits will make it into the dto. It just a question of whether to parse the string when going into the dto or coming out of the dto. I could overload the getters in the dto to parse/return an integer but that is just moving the ugliness somewhere else IMO.

Reply:
Well it would be a better practice to make the DTO reflect the actual datatypes involved.